### ISSUE TITLE
[java] UseTryWithResources false positive when closeable is provided as a method argument or class field

### ISSUE THREAD

**Issue by aaime** [pmd/pmd#3235]
**Affects PMD Version:** 

6.34 (did not happen in 6.29)

**Rule:** [UseTryWithResources](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#usetrywithresources)

**Description:**

``UseTryWithResources`` flags consumer methods getting a ``Closeable`` by argument, with a contract to consume it and close it. However, the ``Closeable`` cannot be put in a try-with-resources, as it's not instatiated in the method itself.
We could debate whether the caller should be entrusted to perform the closing, I'd agree it would be better... but the check should still not report this, a separate check should be developed to enforce this kind of symmetry.

**Code Sample demonstrating the issue:**

Sample code from GeoTools (``FeatureIterator`` is a ``Closeable``):

```java
    @SuppressWarnings("PMD.UseTryWithResources")
    public static int count(FeatureIterator<?> iterator) {
        int count = 0;
        if (iterator != null) {
            try {
                while (iterator.hasNext()) {
                    @SuppressWarnings("unused")
                    Feature feature = iterator.next();
                    count++;
                }
                return count;
            } finally {
                iterator.close();
            }
        }
        return count;
    }
```

**Steps to reproduce:**

Any method of running PMD should produce the issue. In our case, it's run via the maven plugin.

**Comment by oowekyala** [comment]
You *can* use a try with resources though.
```java
try (FeaturedIterator<?> iter = iterator)  {
  ...
}
```
Granted, this aliases the variable, which is not so great for readability. But you don't have to write that finally. 

 In java 9+ you can also directly write `try(iterator) {...}`

I think this is not really a fp, and that it should be up to the user of the rule to decide whether they prefer in this case to close the resource manually (suppress the rule) or alias the closeable with a new variable. Alternatively we could chose to ignore method parameters (unless we're analysing Java 9+ sources). Wdyt?

**Comment by aaime** [comment]
GeoTools is stuck with Java 8 for the short/medium term. I was trying to upgrade from 6.29 to 6.34 but had to stop after a couple of modules (out of 120-ish) because I already had to add a couple dozen suppressions, it was un-manageable :-D .

Ignoring method parameters unless the source is Java 9+ seems like a good approach to me. Aliasing the variable words but looks odd.

Btw, is UseTryWithResources defined as an Xpath (easily overridable in config file) or a class? It seems it's the latter, but cannot find the class in question.

**Comment by aaime** [comment]
> Just for this rule you mean? I'm not sure what changed in this rule between 6.29 and 6.34 that would cause new positives

Correct, just for this new one. I was suprised as well, looked for a possible cause in PMD code changes and could not find one.
But, it's not all bad, it's also catching legit cases that PMD 6.29 was not recognizing (for that same rule).
 
> > Btw, is UseTryWithResources defined as an Xpath (easily overridable in config file) or a class? It seems it's the latter, but cannot find the class in question.
> 
> It's an XPath rule. In our rule docs there's a section with the XPath expression (see [UseTryWithResources](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#usetrywithresources)). For java rules there is instead a link to the source file, see eg [UseCollectionIsEmpty](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#usecollectionisempty). But please, don't override the xpath property in your ruleset. Prefer using [`violationSuppressXPath`](https://pmd.github.io/latest/pmd_userdocs_suppressing_warnings.html#the-property-violationsuppressxpath), or reporting issues to us, because overriding `xpath` will be forbidden in pmd 7. This was cursorily discussed in #2270 if you want more info

Good to know, thanks!

**Comment by aaime** [comment]
I have tried to use "violationSuppressPath", but I'm not sure how it can work. [Documentation](https://pmd.github.io/latest/pmd_userdocs_suppressing_warnings.html#the-property-violationsuppressxpath) says it provides the violation node as the contex node, which would be the try statement. But how I recognize that the variable in question is a method argument?

I've tried to toy with the designer, using this java file, which I though would trigger the rule:

``
import java.util.Iterator;

class Scratch {

    static interface Feature {};

    static interface FeatureIterator extends AutoCloseable, Iterator<Feature> {};

    public static int count(FeatureIterator iterator) {
        int count = 0;
        if (iterator != null) {
            try {
                while (iterator.hasNext()) {
                    iterator.next();
                    count++;
                }
                return count;
            } finally {
                iterator.close();
            }
        }
        return count;
    }
}
``

and setup the rule as:

``xml
//TryStatement[FinallyStatement//Name[
    tokenize(@Image, '\.')[last()] = $closeMethods
][
    pmd-java:typeIs('java.lang.AutoCloseable')
    or
    ../../PrimarySuffix/Arguments//PrimaryPrefix[pmd-java:typeIs('java.lang.AutoCloseable')]
]]
``

but it does not even match... let alone let me try out the suppression:

![image](https://user-images.githubusercontent.com/325433/117535399-d9bfb500-aff5-11eb-980e-c60057a90ffc.png)

I'm clearly doing something wrong, but don't see what :-(

**Comment by oowekyala** [comment]
It doesn't match because of two things:
* the property closeMethods has type `List[String]` and you selected `String` in the editor popup. Also this popup of the designer only supports the default delimiter for lists (`|`). So the value should be `close|closeQuietly`, without spaces.
* even then, you would need to compile your Scratch.java file and place the classes in a jar in the analysis classpath (menu View > Analysis classpath). This is because pmd 6 requires a class file for FeatureIterator to determine that it is Autocloseable. Prefer just using java classes that are already on the classpath. For instance the following example is minimal and does not depend on custom classes:
```java
class Scratch {

    public static int count(AutoCloseable iterator) {
        int count = 0;
        try {
            count++;
        } finally {
            iterator.close();
        }
        return count;
    }
}
```

![Capture d’écran de 2021-05-08 13-47-56](https://user-images.githubusercontent.com/24524930/117538026-86089800-b004-11eb-9dd4-57ebc43afdf9.png)

**Comment by aaime** [comment]
Brilliant... so my "evil plan" was to get back to the variable name, and check if it's among the method argument names, and in that case, skip the check. But, it does not seem to work, because in order to get the variable name, I need tokenize, which is xpath 2.

Trying to explain better, I'm given as the context the ``try`` declaration, so I replicate the xpath of the check to get to the variable name, and then compare to the argument names:

```xml
  <rule ref="category/java/bestpractices.xml/UseTryWithResources" >
    <properties>
      // The rule reports false positives for closeable arguments that are used
      // closed within the method body. This suppression tries to avoid false positives for
      // this case, until we can switch to Java 9 and just use try(variable) {...} without
      // the need to instatiate the variable in the try
      <property name="violationSuppressXPath" value="./FinallyStatement//Name[tokenize(@Image, '\.')[last()] = $closeMethods][pmd-java:typeIs('java.lang.AutoCloseable') and fn:substring-before(@Image, '.') = ancestor::MethodDeclaration/MethodDeclarator//VariableDeclaratorId/@Image]"/>
    </properties>
  </rule>
```

The first part replicates the xpath of the TryWithResources check, in order to get to the very same variable name that triggered the check to fail, then split that name on the dot to get the variable name, and compare using the ancestors with the arguments names. Issues:
-  It's fragile, because in theory the "name" matched could be a chain of calls, but that's unusual, so I'm not worrying too much about it. 
- Uses tokenize, which is not supported by xpath 1
- Uses $closeMethods, which is not provided to the violationSuppressXPath evaluator

So I went with substring and added explicit names rather than using the property... which seems to be working:

```xml
  <rule ref="category/java/bestpractices.xml/UseTryWithResources" >
    <properties>
      // The rule reports false positives for closeable arguments that are used
      // closed within the method body. This suppression tries to avoid false positives for
      // this case, until we can switch to Java 9 and just use try(variable) {...} without
      // the need to instatiate the variable in the try
      <property name="violationSuppressXPath" value="./FinallyStatement//Name[substring-after(@Image, '.') = 'close' or substring-after(@Image, '.') = 'closeQuietly'][pmd-java:typeIs('java.lang.AutoCloseable') and fn:substring-before(@Image, '.') = ancestor::MethodDeclaration/MethodDeclarator//VariableDeclaratorId/@Image]"/>
    </properties>
  </rule>
```

Hum... what do you think? Any obvious mistake?

**Comment by oowekyala** [comment]
This looks fine I guess... at least I don't know how to improve it. I think I'll fix this for the next version so you don't have this in your ruleset for too long.

**Comment by aaime** [comment]
Found false positives also in classes that hold other closeable resources, and have to do something before closing them, but want to make sure they get closed.
Making up an example here:

```java
public class Holder implements AutoCloseable {

   private Reader reader;

   public void close() throws IOException {
      try {
         someOtherActivity();
      } finally {
         reader.close();
      }
   }

   private void someOtherActivity() throws IOException {
     // do stuff
   }
}

**Comment by aaime** [comment]
Another case I've found that's causing false positives is an auto-closeable that happens to be needed in the catch clause, something like the following, assuming `Transaction` is an `AutoCloseable`.

```java
   Transaction t = store.beginTransaction(); 
   try {
       // do something with the store
      t.commit();
  } catch(Exception e) {
      t.rollback();
  } finally {
     t.close();
  }
```

In this case a simple transformation to try-with-resorces cannot be done... an uglier one would be possible, but would require two nested try.
